lib/livebook/update_check.ex

```diff
@@ -0,0 +1,146 @@
+defmodule Livebook.UpdateCheck do
+  @moduledoc false
+
+  # Periodically checks for available Livebook update.
+
+  use GenServer
+
+  require Logger
+
+  @name __MODULE__
+  @timeout :infinity
+
+  @hour_in_ms 60 * 60 * 1000
+  @day_in_ms 24 * @hour_in_ms
+
+  @doc false
+  def start_link(_opts) do
+    GenServer.start_link(__MODULE__, {}, name: @name)
+  end
+
+  @doc """
+  Returns the latest Livebook version if it's more recent than the
+  current one.
+  """
+  @spec new_version() :: String.t() | nil
+  def new_version() do
+    GenServer.call(@name, :get_new_version, @timeout)
+  end
+
+  @doc """
+  Returns whether the update check is enabled.
+  """
+  @spec enabled?() :: boolean()
+  def enabled?() do
+    GenServer.call(@name, :get_enabled, @timeout)
+  end
+
+  @doc """
+  Sets whether the update check is enabled.
+  """
+  @spec set_enabled(boolean()) :: :ok
+  def set_enabled(enabled) do
+    GenServer.cast(@name, {:set_enabled, enabled})
+  end
+
+  @impl true
+  def init({}) do
+    send(self(), :check)
+
+    {:ok,
+     %{
+       enabled: Livebook.Settings.update_check_enabled?(),
+       new_version: nil,
+       timer_ref: nil
+     }}
+  end
+
+  @impl true
+  def handle_cast({:set_enabled, enabled}, state) do
+    Livebook.Settings.set_update_check_enabled(enabled)
+
+    state = cancel_timer(state)
+
+    if enabled do
+      send(self(), :check)
+    end
+
+    {:noreply, %{state | enabled: enabled}}
+  end
+
+  @impl true
+  def handle_call(:get_enabled, _from, state) do
+    {:reply, state.enabled, state}
+  end
+
+  @impl true
+  def handle_call(:get_new_version, _from, state) do
+    new_version = if state.enabled, do: state.new_version
+    {:reply, new_version, state}
+  end
+
+  @impl true
+  def handle_info(:check, state) do
+    state =
+      if state.enabled do
```

Comment:
Now with cancel_timer, you can actually guarantee that this function is never called when the state is disabled. You just need to make sure to not send to yourself on `init`.